source/common/filter/filter_config_discovery_impl.h

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+#include <unordered_map>
+#include <unordered_set>
+
+#include "envoy/config/core/v3/extension.pb.h"
+#include "envoy/config/core/v3/extension.pb.validate.h"
+#include "envoy/config/subscription.h"
+#include "envoy/filter/filter_config_provider.h"
+#include "envoy/protobuf/message_validator.h"
+#include "envoy/server/factory_context.h"
+#include "envoy/singleton/instance.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats_macros.h"
+
+#include "common/config/subscription_base.h"
+#include "common/init/manager_impl.h"
+#include "common/init/target_impl.h"
+
+namespace Envoy {
+namespace Filter {
+
+class HttpFilterConfigProviderManagerImpl;
+class HttpFilterConfigSubscription;
+
+using HttpFilterConfigSubscriptionSharedPtr = std::shared_ptr<HttpFilterConfigSubscription>;
+
+/**
+ * Implementation of a filter config provider using discovery subscriptions.
+ **/
+class DynamicFilterConfigProviderImpl : public HttpFilterConfigProvider {
+public:
+  DynamicFilterConfigProviderImpl(HttpFilterConfigSubscriptionSharedPtr&& subscription,
+                                  const std::set<std::string>& require_type_urls,
+                                  Server::Configuration::FactoryContext& factory_context);
+  ~DynamicFilterConfigProviderImpl() override;
+
+  // Config::ExtensionConfigProvider
+  const std::string& name() override;
+  absl::optional<Http::FilterFactoryCb> config() override;
+  void validateConfig(const ProtobufWkt::Any& proto_config,
+                      Server::Configuration::NamedHttpFilterConfigFactory&) override;
+  void onConfigUpdate(Http::FilterFactoryCb config, const std::string&) override;
+
+private:
+  struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalConfig() : config_{absl::nullopt} {}
+    absl::optional<Http::FilterFactoryCb> config_{};
+  };
+
+  HttpFilterConfigSubscriptionSharedPtr subscription_;
+  const std::set<std::string> require_type_urls_;
+  ThreadLocal::SlotPtr tls_;
+
+  // Local initialization target to ensure that the subscription starts in
+  // case no warming is requested by any other filter config provider.
+  Init::TargetImpl init_target_;
+
+  friend class HttpFilterConfigProviderManagerImpl;
+};
+
+/**
+ * All filter config discovery stats. @see stats_macros.h
+ */
+#define ALL_FILTER_CONFIG_DISCOVERY_STATS(COUNTER)                                                 \
+  COUNTER(config_reload)                                                                           \
+  COUNTER(config_fail)
```

Comment:
These stats should be documented somewhere, also generically s/filter/extension?

Reply:
Agree, will try to find a place to document them.